### PR TITLE
Use parallel stream instead of sequential for loop for pushing writes to table

### PR DESCRIPTION
**Pull request by sushanb**
Replaced with parallelStream fortable writes and uses the atomic boolean to track failures.

### CODE REVIEW

**Comment by sushanb** [comment]
Added tests

File: bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpoint.java [hunk 1]
```diff
@@ -165,25 +165,23 @@ public boolean replicate(ReplicateContext replicateContext) {
...
-        final String tableName = walEntriesForTable.getKey();
-        // TODO: MAKE IT PARALLEL
+      walEntriesByTable.entrySet().parallelStream().forEach(walEntries -> {
+        final String tableName = walEntries.getKey();
         try {
-          boolean result =
-              replicatorFactory.getReplicator(tableName)
-                  .replicateTable(walEntriesForTable.getValue());
-          if (!result) {
+          boolean res = replicatorFactory.getReplicator(tableName)
```
Comment (vermas2012) [inline bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpoint.java]: result or even better tableReplicatedSuccessfully>

File: bigtable-hbase-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java [hunk 2]
```diff
@@ -204,23 +200,49 @@ public void testDelete() throws IOException, InterruptedException {
...
+
+    // Wait for replication to catch up
+    // TODO Find a better alternative than sleeping? Maybe disable replication or turnoff mini cluster
+    Thread.sleep(2000);
+    //   Waiter.waitFor(CONF, 60000, () -> TestEndpoint.getEntries().size() >= cellNum);
+
+    Table cbtTable = cbtConnection.getTable(TableName.valueOf(TABLE_NAME));
+    Table cbtTable2 = cbtConnection.getTable(TableName.valueOf(TABLE_NAME_2));
+    assertResultRowsInCBT(cbtTable, 10, 0);
+    assertResultRowsInCBT(cbtTable, 10, 0);
```
Comment (vermas2012) [inline bigtable-hbase-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java]: cbtTable2?

File: bigtable-hbase-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java [hunk 3]
```diff
@@ -204,23 +200,49 @@ public void testDelete() throws IOException, InterruptedException {
...
       // Only 1 cell/row added in  HBase
       Assert.assertEquals(1, returnedCells.size());
       // Assert on cell contents
       Assert.assertEquals(0, Bytes.compareTo(
           CellUtil.cloneFamily(returnedCells.get(0)), CF1));
       Assert.assertEquals(0, Bytes.compareTo(CellUtil.cloneQualifier(returnedCells.get(0)), COL_QUALIFIER));
-      Assert.assertEquals(0, Bytes.compareTo(CellUtil.cloneValue(returnedCells.get(0)), (VALUE_PREFIX + i).getBytes()));
-      i++;
+      Assert.assertEquals(0, Bytes.compareTo(CellUtil.cloneValue(returnedCells.get(0)), (VALUE_PREFIX + rowStart).getBytes()));
+      rowStart++;
```
Comment (vermas2012) [inline bigtable-hbase-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java]: let rowStart be the start. Take an iterator int and increment it. It looks cleaner and more readable.

File: bigtable-hbase-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java [hunk 4]
```diff
@@ -204,23 +200,49 @@ public void testDelete() throws IOException, InterruptedException {
...
+    }
+
+    // Wait for replication to catch up
+    // TODO Find a better alternative than sleeping? Maybe disable replication or turnoff mini cluster
+    Thread.sleep(2000);
+    //   Waiter.waitFor(CONF, 60000, () -> TestEndpoint.getEntries().size() >= cellNum);
+
+    Table cbtTable = cbtConnection.getTable(TableName.valueOf(TABLE_NAME));
+    Table cbtTable2 = cbtConnection.getTable(TableName.valueOf(TABLE_NAME_2));
+    assertResultRowsInCBT(cbtTable, 10, 0);
```
Comment (vermas2012) [inline bigtable-hbase-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java:271]: This does not test that the replication happens in parallel? 

We need to test that the replication is happening in parallel. Create a custom TestEndpoint and have some mutexes in it. Or leave hooks in beginning and ending to add some counters to make sure that tables were replicating parallely.

**Review by vermas2012 — CHANGES_REQUESTED** [review]

File: bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpoint.java [hunk 5]
```diff
@@ -165,25 +165,23 @@ public boolean replicate(ReplicateContext replicateContext) {
           entries.stream()
               .collect(groupingBy(entry -> entry.getKey().getTablename().getNameAsString()));
 
-      for (Map.Entry<String, List<WAL.Entry>> walEntriesForTable : walEntriesByTable.entrySet()) {
-        final String tableName = walEntriesForTable.getKey();
-        // TODO: MAKE IT PARALLEL
+      walEntriesByTable.entrySet().parallelStream().forEach(walEntries -> {
```
Comment (vermas2012) [inline bigtable-hbase-replication/src/main/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpoint.java]: We should control the number of threads used in this endpoint explicitly. There can be any number of replciated tables and there can be more than 1 endpoint object, 

We should create a fixed size threadpool and force all the tables to replicate on that threadpool. That way we can limit the concurrency on the region server.

File: bigtable-hbase-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java (hunk 4 above)
Comment (sushanb) [reply bigtable-hbase-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java:271]: I am creating a custom TestEndpoint for this test.

Basically

```class HbaseToCloudBigtableReplicationEndpointForTest extends HbaseToCloudBigtableReplicationEndpoint``` 

The parallel replication is abstracted under `replicate` method under `replicationContext.getEntries`. Can I override the entire `replicate` method for test? Any suggestions.

Comment (vermas2012) [reply bigtable-hbase-replication/src/test/java/com/google/cloud/bigtable/hbase/replication/HbaseToCloudBigtableReplicationEndpointTest.java:271]: You will need to create hooks where you can inject test code. One way to do it maybe to supply TestTableReplciators and using them to book keep. If you replicate a WAL entry with 3 tables, you could then expect the test tableReplicators to increment a static atomic integer and block untill it reaches 3. when it reaches 3 you complete the future and all 3 tableReplicators will be free.

**Pull request closed without merging**